wifiphisher/extensions/deauth.py

```diff
@@ -88,101 +63,130 @@ def get_packet(self, packet):
+        receiver = packet.addr1
+        sender = packet.addr2
+        clients = self._add_clients(sender, receiver, bssid)
+
+        if clients:
+            self._observed_clients.append(clients[0])
+            packets_to_send += clients[1]
+
+        return (channels, packets_to_send)
+
+    def _add_clients(self, sender, receiver, bssid):
+        """
+        Craft and add deauthentication packets based on the sender/receiver
+        to the database
+
+        :param self: A Deauth object
+        :param sender: Address of the sender
+        :param receiver: Address of the receiver
+        :param bssid: Address of the bssid
+        :type self: Deauth
+        :type sender: str
+        :type receiver: str
+        :type bssid: str
+        :return: None
+        :rtype: None
+        """
+
+        # list of addresses that are not acceptable
+        non_valid_list = constants.NON_CLIENT_ADDRESSES + self._observed_clients
+
+        # function for passing values
+        pass_fun = lambda s: s
+
+        # check if the sender/receiver is valid
+        is_valid = lambda x: ((x in non_valid_list and pass_fun(False)) or
+                              (x == bssid and pass_fun(True)) or
+                              (pass_fun(False)))
+
+        # craft the packets
+        packets = lambda: (self._craft_packet(receiver, sender, bssid) +
+                           self._craft_packet(sender, receiver, bssid))
+
+        # return the client and packets if valid and None otherwise
+        return (is_valid(sender) and (receiver, packets()) or
+                is_valid(receiver) and (sender, packets()) or
+                None)
 
     def send_output(self):
         """
-        Get all the observed clients.
+        Get any relevant output message
 
-        :param self: A Deauthentication object.
-        :type self: Deauthentication
-        :return: A list with all the Deauth/Disas entries.
+        :param self: A Deauth object
+        :type self: Deauth
+        :return: A list with all the message entries
         :rtype: list
         """
 
-        return ["DEAUTH/DISAS - " + c for c in self._observed_clients]
+        return map(lambda client: "DEAUTH/DISAS - {}".format(client), self._observed_clients)
 
     def send_channels(self):
         """
-        Send channels to subscribe.
+        Send channes to subscribe.
 
-        :param self: A Deauthentication object.
-        :type self: Deauthentication
-        :return: A list with all interested channels.
+        :param self: A Deauth object
+        :type self: Deauth
+        :return: A list with all interested channels
         :rtype: list
         """
-        if self._is_frenzy:
-            return [str(k) for k in constants.ALL_2G_CHANNELS]
-        return [self._data.target_ap_channel]
+
+        channels_to_send = list()
+
+        if self._channels:
+            channels_to_send = self._channels
+            self._channels = list()
+
```

Comment:
The code here looks like not correct for me.

`send_channels` will be called by `em.start_extensions`, and at that time `self._channels` is empty (since you append channel to self._channels in `get_packet`).

We have to initialize channels_to_send first, then it will correctly do the frequecy hopping.